File: insecureapplication/gallery/db/seed.js
// Populates the database with the same fixture data
// insecureapplication/mongo-seed/mongodbdata/gallery2 provided: two OAuth
// clients, one user ("koen"/"password"), two photos, two albums. Idempotent
// (safe to call on every startup) and, like the original mongo-seed
// container, honors CLIENT_ID/CLIENT_SECRET env vars for the PhotoPrint
// client's credentials. Mirrors insecureapplication-go/gallery-idp/db/seed.go.
const fs = require('fs');
const path = require('path');
const bcryptjs = require('bcryptjs');

const clients = require('./clients');
const users = require('./users');
const albumsDb = require('./albums');
const images = require('./images');
const {newID} = require('./id');

// clientId/clientSecret are whatever CLIENT_ID/CLIENT_SECRET currently
// resolve to (default "photoprint"/"secret" if unset). Seeding directly
// under that id -- rather than always seeding "photoprint" and separately
// trying to rename it -- means a restart with a changed CLIENT_ID just
// creates/updates the newly-named client in place; it can't collide with a
// stale row from a *previous* CLIENT_ID, since that row is never touched.
function seedClients(clientId, clientSecret) {
  if (!clients.getClient(clientId)) {
    clients.createClient({clientId, name: 'PhotoPrint', clientSecret, trusted: false});
  } else {
    clients.updateClientFields(clientId, {client_secret: clientSecret});
  }
  if (!clients.getClient('maliciousclient')) {
    clients.createClient({clientId: 'maliciousclient', name: 'maliciousclient', clientSecret: 'secret', trusted: false});
  }
}

function seedUser() {
  let user = users.getUserByUsername('koen');
  if (user) return user._id;
  const hash = bcryptjs.hashSync('password', 10);
  return users.createUser({username: 'koen', name: 'koen', email: 'koen@buyens.org', passwordHash: hash});
}

function seedAlbumsAndPhotos(userId) {
  for (const a of [
    {name: 'Leuven', description: 'Pictures From My University Town'},
    {name: 'default', description: 'Default Album'},
  ]) {
    if (!albumsDb.getAlbumByName(a.name)) {
      albumsDb.createAlbum({name: a.name, description: a.description, userId});
    }
  }

  if (images.listImagesByUser(userId).length > 0) return;
  for (const img of [
    {file: '7ac9eb7f-1de1-4c47-819f-f41591035479.jpg', desc: 'Kuleuven Bib'},
    {file: 'ab1dcbf9-69cd-4fb9-b9e6-ea3a01f992a0.jpg', desc: 'Arenberg Castle'},
  ]) {
    images.createImage({url: img.file, description: img.desc, userId, album: 'Leuven'});
  }
}

// Extracts the embedded seed photos into uploadsDir, skipping any that
// already exist there -- lets a fresh checkout bootstrap its own uploads
// directory. Mirrors gallery-idp's db.WriteSeedUploads.
function writeSeedUploads(uploadsDir) {
  fs.mkdirSync(uploadsDir, {recursive: true});
  const seedDir = path.join(__dirname, 'seedfiles');
  for (const file of fs.readdirSync(seedDir)) {
    const dest = path.join(uploadsDir, file);
    if (fs.existsSync(dest)) continue;
    fs.copyFileSync(path.join(seedDir, file), dest);
  }
}

function seed(uploadsDir) {
  writeSeedUploads(uploadsDir);
  seedClients(process.env.CLIENT_ID || 'photoprint', process.env.CLIENT_SECRET || 'secret');
  const koenID = seedUser();
  seedAlbumsAndPhotos(koenID);
}

module.exports = {seed};

File: insecureapplication-go/gallery-idp/db/seed.go
package db

import (
	"database/sql"
	"fmt"
	"os"

	"golang.org/x/crypto/bcrypt"
)

// Seed populates the database with the same fixture data the original
// insecureapplication/gallery/mongodbdata/gallery2 BSON dump provided:
// two OAuth clients, one user ("koen"/"password"), two photos, two albums.
// It is idempotent (safe to call on every startup) and, like the original
// mongo-seed container, honors CLIENT_ID/CLIENT_SECRET env vars to rewrite
// the "photoprint" client's credentials in place.
func (d *DB) Seed() error {
	if err := d.seedClients(); err != nil {
		return fmt.Errorf("seed clients: %w", err)
	}
	koenID, err := d.seedUser()
	if err != nil {
		return fmt.Errorf("seed user: %w", err)
	}
	if err := d.seedPhotosAndAlbums(koenID); err != nil {
		return fmt.Errorf("seed photos/albums: %w", err)
	}
	return d.applyClientEnvOverride()
}

func (d *DB) seedClients() error {
	targetID := os.Getenv("CLIENT_ID")
	if targetID == "" {
		targetID = "photoprint"
	}
	existing, err := d.GetClient("photoprint")
	if err != nil {
		return err
	}
	if existing == nil && targetID != "photoprint" {
		existing, err = d.GetClient(targetID)
		if err != nil {
			return err
		}
	}
	if existing == nil {
		if err := d.CreateClient("photoprint", "PhotoPrint", "secret", false, ""); err != nil {
			return err
		}
	}
	existing, err = d.GetClient("maliciousclient")
	if err != nil {
		return err
	}
	if existing == nil {
		if err := d.CreateClient("maliciousclient", "maliciousclient", "secret", false, ""); err != nil {
			return err
		}
	}
	return nil
}

func (d *DB) seedUser() (string, error) {
	u, err := d.GetUserByUsername("koen")
	if err != nil {
		return "", err
	}
	if u != nil {
		return u.ID, nil
	}
	hash, err := bcrypt.GenerateFromPassword([]byte("password"), bcrypt.DefaultCost)
	if err != nil {
		return "", err
	}
	id := NewID()
	if err := d.CreateUser(&User{
		ID:           id,
		Username:     "koen",
		Name:         "koen",
		Email:        "koen@buyens.org",
		PasswordHash: string(hash),
	}); err != nil {
		return "", err
	}
	return id, nil
}

func (d *DB) seedPhotosAndAlbums(userID string) error {
	albums := []struct{ name, desc string }{
		{"Leuven", "Pictures From My University Town"},
		{"default", "Default Album"},
	}
	for _, a := range albums {
		existing, err := d.GetAlbumByName(a.name)
		if err != nil {
			return err
		}
		if existing == nil {
			if err := d.CreateAlbum(&Album{ID: NewID(), Name: a.name, Description: a.desc, UserID: sql.NullString{String: userID, Valid: userID != ""}}); err != nil {
				return err
			}
		}
	}

	images, err := d.ListImagesByUser(userID)
	if err != nil {
		return err
	}
	if len(images) > 0 {
		return nil
	}
	seedImages := []struct{ file, desc string }{
		{"7ac9eb7f-1de1-4c47-819f-f41591035479.jpg", "Kuleuven Bib"},
		{"ab1dcbf9-69cd-4fb9-b9e6-ea3a01f992a0.jpg", "Arenberg Castle"},
	}
	for _, img := range seedImages {
		if err := d.CreateImage(&Image{
			ID:          NewID(),
			URL:         img.file,
			Description: img.desc,
			UserID:      userID,
			Album:       "Leuven",
		}); err != nil {
			return err
		}
	}
	return nil
}

// applyClientEnvOverride mirrors mongo-seed/seed.sh's
// `db.clients.updateOne({clientID:'photoprint'}, {$set:{clientID:$CLIENT_ID, clientSecret:$CLIENT_SECRET}}, {upsert:true})`.
func (d *DB) applyClientEnvOverride() error {
	clientID := os.Getenv("CLIENT_ID")
	if clientID == "" {
		clientID = "photoprint"
	}
	clientSecret := os.Getenv("CLIENT_SECRET")
	if clientSecret == "" {
		clientSecret = "secret"
	}
	if clientID == "photoprint" && clientSecret == "secret" {
		return nil // no-op, matches seeded defaults
	}
	existing, err := d.GetClient(clientID)
	if err != nil {
		return err
	}
	if existing != nil {
		if existing.ClientSecret != clientSecret {
			return d.UpdateClientFields(clientID, map[string]string{"client_secret": clientSecret})
		}
		return nil
	}
	return d.UpsertClientCredentials("photoprint", clientID, clientSecret)
}
